### ISSUE TITLE
trivial_bounds doesn't see associated type as implementing a trait

### ISSUE THREAD

**Issue by kornelski** [rust-lang/rust#51044]
Regression in rustc 1.27.0-nightly (2f2a11dfc 2018-05-16), still present in rustc 1.28.0-nightly (29ffe51e3 2018-05-23). Previous day's nightly, and stable, compile fine.

```rust
#![allow(dead_code)]
struct Foo<T>(T);
struct Duck;
struct Quack;

trait Hello<A> where A: Animal {
}

trait Animal {
    type Noise;
}

trait Loud<R>  {
}

impl Loud<Quack> for f32 {
}

impl Animal for Duck {
    type Noise = Quack;
}

impl Hello<Duck> for Foo<f32> where f32: Loud<<Duck as Animal>::Noise> {
}

// impl Hello<Duck> for Foo<f32> where f32: Loud<Quack> // builds fine
```

On rustc 1.27.0-nightly (f0fdaba04 2018-05-15) it builds fine. 
On rustc 1.27.0-nightly (2f2a11dfc 2018-05-16) it fails with:

```
 error[E0277]: the trait bound `f32: Loud<<Duck as Animal>::Noise>` is not satisfied
  --> src/lib.rs:23:1
   |
23 | / impl Hello<Duck> for Foo<f32> where f32: Loud<<Duck as Animal>::Noise> {
24 | | }
   | |_^ the trait `Loud<<Duck as Animal>::Noise>` is not implemented for `f32`
   |
   = help: consider adding a `where f32: Loud<<Duck as Animal>::Noise>` bound
   = help: see issue #48214
   = help: add #![feature(trivial_bounds)] to the crate attributes to enable
```

**Comment by RalfJung** [comment]
Diesel v1.3.0 seems to be affected by this issue, see https://github.com/rust-lang/rust/issues/50825#issuecomment-395746852

**Comment by matthewjasper** [comment]
Fixed by #51042
